File: leetcode/cpp/21.merge-two-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode dummy = ListNode(0);
        ListNode *head = &dummy;
        int a, b;
        while (list1 != nullptr || list2 != nullptr) {
            a = INT_MAX, b = INT_MAX;
            if (list1 != nullptr) {
                a = list1->val;
            }
            if (list2 != nullptr) {
                b = list2->val;
            }
            if (a < b) {
                head->next = new ListNode(a);
                list1 = list1->next;
            } else {
                head->next = new ListNode(b);
                list2 = list2->next;
            }
            head = head->next;
        }
        return dummy.next;
    }
};

/* */

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode* tail = new ListNode(0);
        ListNode* head = tail;
        int l1,l2;
        while(list1 && list2){
            if(list1->val < list2->val){
                tail->next = list1;
                list1 = list1->next;
            }
            else{
                tail->next = list2;
                list2 = list2->next;
            }
            tail = tail->next;
        }
        if(list2){
            tail->next = list2;
        }
        else if(list1){
            tail->next = list1;
        }
        return head->next;
    }
};

File: leetcode/rust-leetcode/src/21.merge-two-sorted-lists.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn merge_two_lists(mut list1: Option<Box<ListNode>>, mut list2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut dummy = Some(Box::new(ListNode::new(0)));
        let mut head = dummy.as_mut();
        while list1.is_some() && list2.is_some() {
            if list1.as_ref()?.val < list2.as_ref()?.val {
                let next = list1.as_mut()?.next.take();
                head.as_mut()?.next = list1;
                list1 = next;
            } else {
                let next = list2.as_mut()?.next.take();
                head.as_mut()?.next = list2;
                list2 = next;
            }
            head = head?.next.as_mut();
        }
        if list1.is_some() {
            head.as_mut()?.next = list1;
        }
        if list2.is_some() {
            head.as_mut()?.next = list2;
        }
        dummy?.next.take()
    }
}

